Add nodiscard attribute

oold opened the pull request:
The `scope_guard` class should be declared `nodiscard` as there really isn't any good reason to discard returned instances of it.

Commit:
Add nodiscard attribute

ricab commented:
Plus, the caller could still discard if they wanted (e.g. cast, assign to `std::ignore`). 

Thanks @oold! I am just trying to get Github Actions back into shape first.

ricab committed:
[tests] Adapt tests to nodiscard attr

ricab committed:
[cmake] Allow unused-result warnings in old GCCs

Do not treat unused-result warnings as errors in versions of gcc before
10, as they (wrongly) issued that warning in unevaluated contexts. See
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=89070.
